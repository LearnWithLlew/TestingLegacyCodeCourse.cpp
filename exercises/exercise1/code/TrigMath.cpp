#include "TrigMath.h"

double TrigMath::SinRaw(int idx) {
    return SIN_TABLE[idx & SIN_MASK];
}

double TrigMath::CosRaw(int idx) {
    return SIN_TABLE[(idx + COS_OFFSET) & SIN_MASK];
}

double TrigMath::Mxatan(double arg) {
    auto argsq = arg * arg;
    auto value = (((p4 * argsq + p3) * argsq + p2) * argsq + p1) * argsq + p0;
    value /= ((((argsq + q4) * argsq + q3) * argsq + q2) * argsq + q1) * argsq + q0;
    return value * arg;
}

double TrigMath::Msatan(double arg) {
    if (arg < sq2m1)
        return Mxatan(arg);
    if (arg > sq2p1)
        return HALF_PI - Mxatan(1 / arg);
    return HALF_PI / 2 + Mxatan((arg - 1) / (arg + 1));
}

void TrigMath::CreateSinTable() {
    for (auto i = 0; i < SIN_SIZE; i++)
        SIN_TABLE[i] = sin(i * TWO_PI / SIN_SIZE);
}

TrigMath::TrigMath() {
    CreateSinTable();
}

double TrigMath::Sin(double angle) {
    return SinRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

int TrigMath::Floor(double a) {
    return (int) a;
}

double TrigMath::Cos(double angle) {
    return CosRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

double TrigMath::Tan(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return SinRaw(idx) / CosRaw(idx);
}

double TrigMath::Sec(double angle) {
    return 1 / Cos(angle);
}

double TrigMath::Cot(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return CosRaw(idx) / SinRaw(idx);
}

double TrigMath::Asin(double value) {
    if (value > 1)
        return NAN;
    if (value < 0)
        return -Asin(-value);
    auto temp = sqrt(1 - value * value);
    if (value > 0.7)
        return HALF_PI - Msatan(temp / value);
    return Msatan(value / temp);
}

double TrigMath::Acos(double value) {
    if (value > 1 || value < -1)
        return NAN;
    return HALF_PI - Asin(value);
}

double TrigMath::Atan(double value) {
    if (value > 0)
        return Msatan(value);
    return -Msatan(-value);
}

double TrigMath::Atan2(double y, double x) {
    if (y + x == y)
        return y >= 0 ? HALF_PI : -HALF_PI;
    y = Atan(y / x);
    if (x < 0) {
        if (y <= 0) {
            return y + PI;
        } else {
            return y - PI;
        }
    }
    return y;
}

double TrigMath::Acsc(double value) {
    if (value == 0)
        return NAN;
    return Asin(1 / value);
}

double TrigMath::Asec(double value) {
    if (value == 0)
        return NAN;
    return Acos(1 / value);
}

double TrigMath::Acot(double value) {
    if (value == 0)
        return NAN;
    if (value > 0)
        return Atan(1 / value);
    return Atan(1 / value) + PI;
}

