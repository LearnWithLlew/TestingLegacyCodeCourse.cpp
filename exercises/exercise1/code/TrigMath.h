//
// Created by LLEWELLYN FALCO on 12/10/17.
//

#ifndef CATCHPLAYGROUND_TRIGMATH_H
#define CATCHPLAYGROUND_TRIGMATH_H

#define _USE_MATH_DEFINES
#include <math.h>
#include <cmath>

class TrigMath {
public:
    const double PI = M_PI;
    const double SQUARED_PI = PI * PI;
    const double HALF_PI = PI / 2;
    const double QUARTER_PI = HALF_PI / 2;
    const double TWO_PI = 2 * PI;
    const double THREE_PI_HALVES = TWO_PI - HALF_PI;
    const double DEG_TO_RAD = PI / 180;
    const double HALF_DEG_TO_RAD = PI / 360;
    const double RAD_TO_DEG = 180 / PI;
    double SQRT_OF_TWO = std::sqrt(2);
    double HALF_SQRT_OF_TWO = SQRT_OF_TWO / 2;
private  :

    const int SIN_SIZE = 100000;
    double SIN_TABLE[100000];
    const int SIN_MASK = SIN_SIZE - 1;
    const double SIN_CONVERSION_FACTOR = SIN_SIZE / TWO_PI;

    const int COS_OFFSET = SIN_SIZE / 4;

//Arc trig
    const double sq2p1 = 2.414213562373095048802;

    const double sq2m1 = 0.414213562373095048802;
    const double p4 = 0.161536412982230228262E2;
    const double p3 = 0.26842548195503973794141E3;
    const double p2 = 0.11530293515404850115428136E4;
    const double p1 = 0.178040631643319697105464587E4;
    const double p0 = 0.89678597403663861959987488E3;
    const double q4 = 0.5895697050844462222791E2;
    const double q3 = 0.536265374031215315104235E3;
    const double q2 = 0.16667838148816337184521798E4;
    const double q1 = 0.207933497444540981287275926E4;
    const double q0 = 0.89678597403663861962481162E3;

    double sinRaw(int idx);

    double cosRaw(int idx);

    double mxatan(double arg);

    double msatan(double arg);

    void createSinTable();

public:

    TrigMath();


    double sin(double angle);

    int floor(double a);


    double cos(double angle);

    double tan(double angle);

    double csc(double angle) {
        return 1 / sin(angle);
    }

    double sec(double angle);


    double cot(double angle);

    double asin(double value);

    double acos(double value);

    double atan(double value);

    double atan2(double y, double x);

    double acsc(double value);

    double asec(double value);

    double acot(double value);


};

#endif //CATCHPLAYGROUND_TRIGMATH_H
