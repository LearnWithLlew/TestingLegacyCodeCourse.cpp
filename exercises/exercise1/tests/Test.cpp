#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"
#include "../code/TrigMath.h"

#include <sstream>

using namespace ApprovalTests;

namespace
{
    void runOne(std::ostream& s, std::string function, double input, double result)
    {
        s << function << input << ": " << result << '\n';
    }
    std::string runEverything(double value)
    {
        TrigMath math;
        std::stringstream s;
        s << "input value: " << value << '\n';
        runOne(s, "Sin()  : ", value, math.Sin(value));
        runOne(s, "Floor(): ", value, math.Floor(value) );
        runOne(s, "Cos()  : ", value, math.Cos(value) );
        runOne(s, "Tan()  : ", value, math.Tan(value) );
        runOne(s, "Csc()  : ", value, math.Csc(value) );
        runOne(s, "Sec()  : ", value, math.Sec(value) );
        runOne(s, "Cot()  : ", value, math.Cot(value) );
        runOne(s, "Asin() : ", value, math.Asin(value) );
        runOne(s, "Acos() : ", value, math.Acos(value) );
        runOne(s, "Atan() : ", value, math.Atan(value) );
        runOne(s, "Atan2(): ", value, math.Atan2(value, value) );
        runOne(s, "Acsc() : ", value, math.Acsc(value) );
        runOne(s, "Asec() : ", value, math.Asec(value) );
        runOne(s, "Acot() : ", value, math.Acot(value) );
        return s.str();
    }
}

TEST_CASE("Test Sin") {
    TrigMath math;
    REQUIRE(Approx(0.8676960543) == math.Sin(3.4));
}

TEST_CASE("Test Asin") {
    TrigMath math;
    REQUIRE(Approx(-0.7753974966) == math.Asin(-0.7));
    REQUIRE(Approx(0.304692654) == math.Asin(0.3));
    REQUIRE(Approx(1.119769515) == math.Asin(0.9));
    REQUIRE(isnan(math.Asin(1.2)));
}

TEST_CASE("Approve Sin")
{
    TrigMath math;
    Approvals::verify(math.Sin(3.2));
}

TEST_CASE("Approve Cos")
{
    TrigMath math;
    Approvals::verify(math.Cos(3.2));
}

TEST_CASE("Approve everything - for 1.0")
{
    Approvals::verify(runEverything(1.0));
}

TEST_CASE("Approve everything - for 3.14")
{
    Approvals::verify(runEverything(3.14));
}

TEST_CASE("Approve everything - for 1.0 and 3.14")
{
    std::vector<double> values{ 1.0, 3.14 };
    Approvals::verifyAll("runEverything", values, [](double value, std::ostream& o) {
        o << runEverything(value);
        }
    );
}

TEST_CASE("Combination approvals - for one input")
{
    std::vector<double> values{ 1.0 };
    CombinationApprovals::verifyAllCombinations(
        [](double value) { return runEverything(value); }, values);
}

TEST_CASE("Approve Atan2 - all values")
{
    std::vector<double> values{ -0.5, 0, 0.5, 1.0 };
    TrigMath math;
    CombinationApprovals::verifyAllCombinations(
        [&math](double value1, double value2) { return math.Atan2(value1, value2); },
        values, values);
}

