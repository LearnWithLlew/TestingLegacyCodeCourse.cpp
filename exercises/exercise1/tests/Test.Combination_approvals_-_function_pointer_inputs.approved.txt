(Acos() : , -0.5) => 2.0944
(Acos() : , 0) => 1.5708
(Acos() : , 0.5) => 1.0472
(Acos() : , 1) => 0
(Acot() : , -0.5) => 2.03444
(Acot() : , 0) => -nan(ind)
(Acot() : , 0.5) => 1.10715
(Acot() : , 1) => 0.785398
(Acsc() : , -0.5) => nan
(Acsc() : , 0) => -nan(ind)
(Acsc() : , 0.5) => -nan(ind)
(Acsc() : , 1) => 1.5708
(Asec() : , -0.5) => -nan(ind)
(Asec() : , 0) => -nan(ind)
(Asec() : , 0.5) => -nan(ind)
(Asec() : , 1) => 0
(Asin() : , -0.5) => -0.523599
(Asin() : , 0) => 0
(Asin() : , 0.5) => 0.523599
(Asin() : , 1) => 1.5708
(Atan() : , -0.5) => -0.463648
(Atan() : , 0) => 0
(Atan() : , 0.5) => 0.463648
(Atan() : , 1) => 0.785398
(Cos()  : , -0.5) => 0.0413944
(Cos()  : , 0) => 0.00854503
(Cos()  : , 0.5) => 0.878567
(Cos()  : , 1) => 0.828807
(Cot()  : , -0.5) => -0.423805
(Cot()  : , 0) => inf
(Cot()  : , 0.5) => 8.99496
(Cot()  : , 1) => 8.54019
(Csc()  : , -0.5) => -10.2382
(Csc()  : , 0) => inf
(Csc()  : , 0.5) => 10.2382
(Csc()  : , 1) => 10.3042
(Sec()  : , -0.5) => 24.1579
(Sec()  : , 0) => 117.027
(Sec()  : , 0.5) => 1.13822
(Sec()  : , 1) => 1.20655
(Sin, -0.5) => -0.0976732
(Sin, 0) => 0
(Sin, 0.5) => 0.0976732
(Sin, 1) => 0.0970479
(Sin()  : , -0.5) => -0.0976732
(Sin()  : , 0) => 0
(Sin()  : , 0.5) => 0.0976732
(Sin()  : , 1) => 0.0970479
(Tan()  : , -0.5) => -2.35958
(Tan()  : , 0) => 0
(Tan()  : , 0.5) => 0.111173
(Tan()  : , 1) => 0.117093

